### PR TITLE
Render mobileapp ads

### PR DESCRIPTION
**Pull request by jaiminpanchal27**

### CODE REVIEW

File: src/creative.js [hunk 1]
```diff
@@ -120,35 +133,96 @@ function renderCrossDomain(adId, pubUrl) {
...
-    if (bidObject.adm && bidObject.nurl) {
-      ad = bidObject.adm;
-      ad += utils.createTrackPixelHtml(decodeURIComponent(bidObject.nurl));
-      utils.writeAdHtml(ad);
-    } else if (bidObject.adm) {
-      ad = bidObject.adm;
+    let width = (bidObject.width) ? bidObject.width : bidObject.w;
+    let height = (bidObject.height) ? bidObject.height : bidObject.h;
+    if (bidObject.adm) {
+      ad = (!isMobileApp) ? bidObject.adm : constructMarkup(bidObject.adm, width, height);
```
Comment (mkendall07) [inline src/creative.js]: can you swap this logic (don't use !) ? It's confusing to read written this way.

File: src/creative.js [hunk 2]
```diff
@@ -120,35 +133,96 @@ function renderCrossDomain(adId, pubUrl) {
...
+      if (bidObject.nurl) {
+        ad += utils.createTrackPixelHtml(decodeURIComponent(bidObject.nurl));
+      }
       utils.writeAdHtml(ad);
     } else if (bidObject.nurl) {
-      let nurl = bidObject.nurl;
-      utils.writeAdUrl(nurl, bidObject.h, bidObject.w);
+      if(!isMobileApp) {
```
Comment (mkendall07) [inline src/creative.js]: don't need ! here either

File: src/utils.js [hunk 3]
```diff
@@ -61,3 +61,48 @@ export function insertElement(elm, doc, target) {
...
+  });
+  return uuid;
+};
+
+export function loadScript(currentWindow, tagSrc, callback) {
+  let doc = currentWindow.document;
+  let scriptTag = doc.createElement('script');
+  scriptTag.type = 'text/javascript';
+
+  // Execute a callback if necessary
```
Comment (mkendall07) [inline src/utils.js]: this is dead code

File: src/creative.js [hunk 4]
```diff
@@ -120,35 +133,96 @@ function renderCrossDomain(adId, pubUrl) {
...
+        let nurl = bidObject.nurl;
+        utils.writeAdUrl(nurl, width, height);
+      } else {
+        let adhtml = utils.loadScript(window, bidObject.nurl);
+        ad = constructMarkup(adhtml.outerHTML, width, height);
```
Comment (mkendall07) [inline src/creative.js]: interesting approach!

File: src/creative.js [hunk 5]
```diff
@@ -120,35 +133,96 @@ function renderCrossDomain(adId, pubUrl) {
...
-      bidObject = JSON.parse(response);
-    } catch (error) {
-      console.log(`Error parsing response from cache host: ${error}`);
-    }
-    
+/**
+ * Cache request Callback to display creative
+ * @param {Bool} isMobileApp 
+ */
+function handler(isMobileApp) {
```
Comment (mkendall07) [inline src/creative.js]: this could have a more descriptive name

File: src/creative.js [hunk 6]
```diff
@@ -27,20 +27,28 @@ const DEFAULT_CACHE_PATH = '/pbc/v1/cache';
...
+ * Public render ad function to be used in dfp creative setup
  * @param  {object} doc
  * @param  {string} adId
  * @param  {dataObject} dataObject
  */
 pbjs.renderAd = function(doc, adId, dataObject) {
-  if (environment.isAmp(dataObject)) {
-    renderAmpAd(dataObject.cacheHost, dataObject.cachePath, dataObject.uuid);
+  if(environment.isMobileApp(dataObject)) {
+    renderAmpOrMobileAd(dataObject.cacheHost, dataObject.cachePath, dataObject.uuid, true);
```
Comment (mkendall07) [inline src/creative.js]: why pass a flag for isMobile? Just create 2 functions.

**Comment by jaiminpanchal27** [comment]
Updated based on above feedback

File: src/creative.js (hunk 5 above)
Comment (jaiminpanchal27) [reply src/creative.js]: Updated name to responseCallback

File: src/creative.js (hunk 6 above)
Comment (jaiminpanchal27) [reply src/creative.js]: Lot of code is common between these two functions so decided to added a flag

**Pull request merged**